From 7892d979813c550df04ed84c537df2b5864eb6d5 Mon Sep 17 00:00:00 2001
From: Hanlu Li <heuleehanlu@gmail.com>
Date: Mon, 21 Apr 2025 17:31:34 +0800
Subject: [PATCH 1058/1058] LATX, fix: Fix ioctl command mismatch due to high
 16 bits

This patch changes the behavior to log a warning when a mismatch is detected,
but still proceeds with handing the ioctl. This improves compatibility with
headers that use different high bits, such as amdgpu_drm.h.

Signed-off-by: Hanlu Li <heuleehanlu@gmail.com>
---
 linux-user/syscall.c | 14 +++++++++++---
 1 file changed, 11 insertions(+), 3 deletions(-)

diff --git a/linux-user/syscall.c b/linux-user/syscall.c
index 9008e7adad..3e1fa78218 100644
--- a/linux-user/syscall.c
+++ b/linux-user/syscall.c
@@ -8159,13 +8159,21 @@ static abi_long do_ioctl(int fd, int cmd, abi_ulong arg)
     ie += ioctl_cache_val;
 
 #ifdef TARGET_X86_64
-    if (ioctl_cache_unsupported_cmd[hash] == cmd) {
+    if ((uint16_t)ioctl_cache_unsupported_cmd[hash] == (uint16_t)cmd) {
         ret = get_errno(safe_ioctl(fd, cmd, arg));
         return ret;
     }
 #endif
     /* slow path to iterate the ioctl_entries */
-    if (ie->target_cmd != cmd) {
+    if ((uint16_t)ie->target_cmd != (uint16_t)cmd) {
+#ifdef CONFIG_LATX_DEBUG
+        if (ie->target_cmd != cmd) {
+            qemu_log_mask(LAT_LOG_SYSCALL,
+                    "[LATX_SYSCALL] do_ioctl fd " TARGET_FMT_ld
+                    " cmd 0x" TARGET_FMT_lx " ie->target_cmd != cmd\n",
+                    arg1, arg2);
+        }
+#endif
         switch(TARGET_IOC_NR(cmd)) {
             case TARGET_IOC_NR(TARGET_HIDIOCSFEATURE(0)):
             case TARGET_IOC_NR(TARGET_HIDIOCGFEATURE(0)):
@@ -8190,7 +8198,7 @@ static abi_long do_ioctl(int fd, int cmd, abi_ulong arg)
                 return -TARGET_ENOSYS;
 #endif
             }
-            if (ie->target_cmd == cmd) {
+            if ((uint16_t)ie->target_cmd == (uint16_t)cmd) {
                 ioctl_cache[hash] = i;
 #ifdef TARGET_X86_64
                 ioctl_cache_unsupported_cmd[hash] = 0;
-- 
2.49.0

